### ISSUE TITLE
First login to harbor with oidc-auth and api/curl

### ISSUE THREAD

**Issue by dm0610** [goharbor/harbor#16047]
Hello everyone.
Tell me, is there a real opportunity to make the first login to Harbor via oidс not through GUI, but through api? 
I easily get the id_token from the keycloak, but I can't log in. 

We are trying to automate getting of cli-tokens for technical accounts.
here i get token
`    private static String auth(script, String credentialsId){
        script.withCredentials([script.usernamePassword(credentialsId: credentialsId,
            usernameVariable: 'USER', passwordVariable: 'PASS'),
            script.usernamePassword(credentialsId: 'admin_user_id', usernameVariable: 'USER2', passwordVariable: 'PASS2')]) {
            def out = script.sh script: "curl -s --location --request POST 'https://sso.mycompany.com/auth/realms/my_company/protocol/openid-connect/token' \
            --header 'Content-Type: application/x-www-form-urlencoded' \
            --data-urlencode 'grant_type=password' \
            --data-urlencode 'client_id=$script.USER2' \
            --data-urlencode 'client_secret=$script.PASS2' \
            --data-urlencode 'scope=openid' \
            --data-urlencode 'response_type=id_token' \
            --data-urlencode 'username=$script.USER' \
            --data-urlencode 'password=$script.PASS'", returnStdout: true
            return script.readJSON(text: out).id_token
        }
    }`
After that I try to connect via api:
curl -v -H "Authorization:  Bearer $token" https://harbor.mycompany.com/api/v2.0/users/current
`{ [5 bytes data]
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Server: nginx
< Date: Tue, 23 Nov 2021 13:39:42 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 62
< Connection: keep-alive
< Set-Cookie: sid=55abaabc7eaef26vvqc84f02423e5ca45e; Path=/; Secure; HttpOnly
< X-Request-Id: 7f72qa2ab-4422-45bf-a455-70dfs26f3e9732
< 
{ [62 bytes data]
{"errors":[{"code":"UNAUTHORIZED","message":"unauthorized"}]}`

Is there a real solution to this problem?

**Comment by Vad1mo** [comment]
- A user can get the CLI Token with that he can do also API Calls and push/pull images with docker.
- Another approach is to use robot accounts, push pull operation. 
 
 There is no other way. 
 
 Could you explain what you are trying to achieve a bit more in detail or why it's difficult for a user to login to Harbor and get his token?

**Comment by dm0610** [comment]
@Vad1mo hello, Thanks for the answer.
I know about robot accounts. Their main fault is that they do not work with harbor api. Therefore, our team decided to use domain technical accounts. A keyloak is used for authentication. One problem is that in Harbor, for authentication using OIDC, you must log in for the first time through the browser (GUI). There is no such possibility for api. Therefore, any automation for creating cli tokens loses its meaning - you will have to manually login first ((
Hence my question - maybe I just did not find a way to make the first login via api

**Comment by Vad1mo** [comment]
here is a hack: how you might use robot accounts for api calls:
1. Enable Debugging on Harbor Core
2. Create a system-wide robot account (the permission set does not matter)
3. run the API operation 
4. Check logs what permission is missing
5. Add the missing permission in the DB manually
6. Repeat until you don't see any errors!

**Comment by mymarche** [comment]
Hack it's a nice, but why don't realize the https://auth0.com/docs/authorization/flows/client-credentials-flow

**Comment by sagikazarmark** [comment]
This is really a pain for a lot of reasons:

- The documentation is absolutely useless, the wiki contains some random links to issues with no real details, so people will have to find most if this out for themselves.
- The average user can't be expected to "add missing permissions in the DB manually"
- Similarly, the general direction to "get an ID token somehow" isn't something that an average user will do.

Robot accounts can't do everything (without some manual hacking), so using robot accounts for API authentication is not always going to be enough.

Am I missing something? After spending half a day in the docs and on the issue tracker, this is what I managed to figure out.

**Comment by chazapis** [comment]
I am facing a similar issue (also mentioned in #15556, #13683, and many others...). I need to get a user's id and CLI token *before* the user logs in for the first time "properly", i.e. via a browser. Since Harbor is configured to use OIDC, there is no way to create a user programmatically.

An easy fix would be to allow just that. Now, when POSTing to `/users`, I get 403, since OIDC is enabled. I suggest the admin to be allowed to create a user with a username (and nothing else) via the API, and Harbor to assign an id and CLI token. When the actual user does eventually log in via a browser, he/she should be mapped to the existing account with the same username (which I guess is already done).

As a workaround, I now have to monitor logins via the OIDC server and trigger my code several seconds after a user logs in to Harbor, which is impractical.

**Comment by prashi2202** [comment]
@Vad1mo : A user can get the CLI Token with that he can do also API Calls and push/pull images with docker. 

How do we do that ? My setup is behind Okta OIDC provider and Harbor running on EKS cluster.

**Comment by sagikazarmark** [comment]
Bump

**Comment by chazapis** [comment]
Bump!

**Comment by mymarche** [comment]
Bump

**Comment by chazapis** [comment]
Bump. Please see my comment above on an easy fix for this.

**Comment by abdulsalamalmahdi** [comment]
Bump

**Comment by reasonerjt** [comment]
> An easy fix would be to allow just that. Now, when POSTing to /users, I get 403, since OIDC is enabled. I suggest the admin to be allowed to create a user with a username (and nothing else) via the API, and Harbor to assign an id and CLI token. When the actual user does eventually log in via a browser, he/she should be mapped to the existing account with the same username (which I guess is already done).

I don't think this flow would work.  When the user is added by admin, it's not yet verified by the Identity provider, how come he can have a CLI token?

**Comment by ajoga** [comment]
Hey, I'm new to Harbor, there is an "[Automatic onboarding](https://goharbor.io/docs/2.1.0/administration/configure-authentication/oidc-auth/)" option that I enabled, the documentation states:

> If the option Automatic onboarding is checked, you will not be prompted to set the user name, instead, Harbor will try to extract the user name from ID token via the claim set in Username Claim and automatically onboard the user using this username.

Is there still a rationale to have that first browser-based interaction at all?
